### PR TITLE
Ensure that tab characters aren't in multi-line strings before throwing a violation

### LINKED ISSUE #3674: W191 False positive for multiline string
Applying ruff on the following file (notice the tab indent in the string), ruff yields a error:
a.py :
```python
a="""
	a
"""
```
This is the command I executed:
```bash
$ ruff --version
ruff 0.0.257
$ ruff a.py --select W191
a.py:2:1: W191 Indentation contains tabs
Found 1 error.
```
Ruff should leave alone the text inside strings, since it could be anything.

### PR DESCRIPTION
**Pull request by evanrittenhouse**
Fixes #3674 and #3438

### CODE REVIEW

File: crates/ruff/src/checkers/physical_lines.rs [hunk 1]
```diff
@@ -154,9 +156,24 @@ pub fn check_physical_lines(
             }
         }
 
+        // Find the quote delimiter in the line before checking the indentation
+        let (quote_delimiter, occurrences) = find_delimiter(line);
+        if quote_delimiter.is_some() {
+            in_quote = !in_quote;
+        }
```
Comment (charliermarsh) [inline crates/ruff/src/checkers/physical_lines.rs:163]: Hmm, I think it's going to be difficult to make this sufficiently performant and robust with the current approach.

One way to think about what's happening here is that we already go through this process of identifying strings when we _tokenize_ the source code, so we're repeating a lot of work here.

Another observation is that we're paying a high cost in the common case to avoid an edge case (most users won't have this rule enabled, most code won't contain tabs, and many tabs won't be in strings).

A few ideas that we could try to leverage here:

1. We could add a `Vec<Range>` to `Indexer` to track all known string locations. Then here, if we detect tab indentation, we could check whether the location is within one of those string ranges. Would need to benchmark to know whether this is sufficiently performant. We already iterate over all tokens there, so grabbing the string ranges is cheap, but we do then have to store them all. (We'd also only need to check "Is this location in a string?" in the event that we detect a tab, so we wouldn't have to pay that price continuously as we do here.)
2. Maybe @MichaReiser has an idea of how this could be achieved within the logical lines framework?

Comment (evanrittenhouse) [reply crates/ruff/src/checkers/physical_lines.rs:163]: That's a good point @charliermarsh, I didn't think about the tokenization process. Where do we do that? I like your Vec<Range> idea. That would be much cleaner.

Comment (charliermarsh) [reply crates/ruff/src/checkers/physical_lines.rs:163]: If you take a look at `crates/ruff_python_ast/src/source_code/indexer.rs`, that struct gets a chance to extract meta-information from the token stream, so you could include `Tok::String` in that loop and build up a list of locations. If you look at the caller of _this_ function, it passes in `indexer.commented_lines()` already. So you could change it to take `indexer` directly.

Comment (evanrittenhouse) [reply crates/ruff/src/checkers/physical_lines.rs:163]: Sounds good. Let me try doing that - hopefully the benchmarks improve with that approach. Doing it this way felt off anyway, but that's on me; I didn't even think to check for tokenization. I need to brush up on the whole parsing/lexing/tokenizing process

Comment (evanrittenhouse) [reply crates/ruff/src/checkers/physical_lines.rs:163]: Marking as draft until that's done.

File: crates/ruff_python_ast/src/source_code/indexer.rs [hunk 2]
```diff
@@ -114,4 +134,73 @@ import os
...
+    #[test]
+    fn string_lines() {
+        let contents = r#""this is a string""#;
+        let lxr: Vec<LexResult> = lexer::lex(contents, Mode::Module).collect();
+        let indexer: Indexer = lxr.as_slice().into();
+        assert!(
+            indexer.string_lines()[0].location.row() == 1
+                && indexer.string_lines()[0].end_location.row() == 1
+                && indexer.string_lines().len() == 1
+        );
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/source_code/indexer.rs:147]: Nit: You can use `assert_eq(indexer.string_lines(), &vec![Range::new(...)])` to do the comparision. This has the advantage that the test prints a nice diff if `string_lines` is empty or contains additional elements.

File: crates/ruff_python_ast/src/source_code/indexer.rs [hunk 3]
```diff
@@ -114,4 +134,73 @@ import os
...
+        let lxr: Vec<LexResult> = lexer::lex(contents, Mode::Module).collect();
+        let indexer: Indexer = lxr.as_slice().into();
+        assert!(
+            indexer.string_lines()[0].location.row() == 2
+                && indexer.string_lines()[0].end_location.row() == 2
+                && indexer.string_lines().len() == 1
+        );
+
+        let contents = r#"
+        """
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/source_code/indexer.rs:187]: Could you add a test for a single line string?

File: crates/ruff_python_ast/src/source_code/indexer.rs [hunk 4]
```diff
@@ -18,17 +21,32 @@ impl Indexer {
...
+            if matches!(
+                tok,
+                Tok::String {
+                    value: _,
+                    kind: _,
+                    triple_quoted: true
+                }
+            ) {
+                string_lines.push(Range::new(*start, *end));
+            }
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/source_code/indexer.rs:49]: Nit:
```suggestion
			match tok {
				Tok::Comment(_) => {
	                commented_lines.push(start.row());
	            }
                Tok::String {
                    value: _,
                    kind: _,
                    triple_quoted: true
                } => {
                	string_lines.push(Range::new(*start, *end));
	            }
```

To express that the branches are mutually exclusive

File: crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs [hunk 5]
```diff
@@ -15,11 +15,53 @@ impl Violation for TabIndentation {
...
+        if start <= lineno && lineno <= end {
+            return Some(curr);
+        } else if start > lineno {
+            high = middle - 1;
+        } else if end < lineno {
+            low = middle + 1;
+        }
+    }
+
+    Some(&string_lines[high])
```
Comment (MichaReiser) [inline crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs:48]: You can use rust's built in [`binary-search`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search) method.

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff_python_ast/src/source_code/indexer.rs (hunk 3 above)
Comment (evanrittenhouse) [reply crates/ruff_python_ast/src/source_code/indexer.rs:187]: Yep, for sure. It also seems like there's a corner case I need to fix. With the current implementation, these two lines ([88](https://github.com/charliermarsh/ruff/blob/main/crates/ruff/resources/test/fixtures/pycodestyle/W19.py#L88), [137](https://github.com/charliermarsh/ruff/blob/main/crates/ruff/resources/test/fixtures/pycodestyle/W19.py#L137)) don't get counted as violations. I'll look into it after work, I'm wondering if it's maybe something where the tab character is on the line after a triple-quoted string.

Comment (evanrittenhouse) [reply crates/ruff_python_ast/src/source_code/indexer.rs:187]: Actually, @MichaReiser I believe the single-line strings are covered on lines 140 and 176 of the file diff. Is that what you meant, or did you mean something like 
```
r""""multiline string on one line""""#;
```

Comment (evanrittenhouse) [reply crates/ruff_python_ast/src/source_code/indexer.rs:187]: Though it also seems like `cargo test` doesn't run `indexer` unit tests. Is there a specific command we have for those?

Comment (MichaReiser) [reply crates/ruff_python_ast/src/source_code/indexer.rs]: Uhm sorry. I meant adding a test for a single quote string (`'test'` or `"test"`) to make sure the indexer doesn't pick them up incorrectly.

Comment (evanrittenhouse) [reply crates/ruff_python_ast/src/source_code/indexer.rs]: No problem! That's covered in the first test

**Comment by charliermarsh** [comment]
(Doing some testing, then will merge tonight!)

File: crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs [hunk 6]
```diff
@@ -16,15 +16,48 @@ impl Violation for TabIndentation {
...
+        if let Ok(range_index) = string_ranges.binary_search_by(|range| {
+            let start = range.location.row();
+            let end = range.end_location.row();
+            if start > lineno {
+                std::cmp::Ordering::Greater
+            } else if end < lineno {
+                std::cmp::Ordering::Less
+            } else {
+                std::cmp::Ordering::Equal
+            }
```
Comment (charliermarsh) [inline crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs:32]: @evanrittenhouse - Does this look reasonable, to use the built-in `binary_search_by` here?

File: crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs [hunk 7]
```diff
@@ -16,15 +16,48 @@ impl Violation for TabIndentation {
...
+
+            // Tab on first line of the quoted range, following the quote.
+            if lineno == start.row() && tab_column > start.column() {
+                return None;
+            }
+
+            // Tab on last line of the quoted range, preceding the quote.
+            if lineno == end.row() && tab_column < end.column() {
+                return None;
+            }
```
Comment (charliermarsh) [inline crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs:53]: I think it's sufficient to just check these, and not check the "Is it on the start _and_ end line" condition, since if it's on the start _and_ end line, it'll be caught by the first condition... But hoping I'm not thinking about it incorrectly.

File: crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs [hunk 8]
```diff
@@ -16,15 +16,48 @@ impl Violation for TabIndentation {
...
+            let string_range = &string_ranges[range_index];
+            let start = string_range.location;
+            let end = string_range.end_location;
+
+            // Tab is contained in the string range by virtue of lines.
+            if lineno != start.row() && lineno != end.row() {
+                return None;
+            }
+
+            let tab_column = line[..tab_index].chars().count();
```
Comment (charliermarsh) [inline crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs:43]: (I think we need to count characters here, since `.column()` is based on characters.)

File: crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs (hunk 6 above)
Comment (evanrittenhouse) [reply crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs:32]: Yep! Wasn't aware it existed, seems much cleaner

File: crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs (hunk 8 above)
Comment (evanrittenhouse) [reply crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs:43]: How do you mean?

File: crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs (hunk 7 above)
Comment (evanrittenhouse) [reply crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs:53]: Nah, I think that's right. IIRC I included a test case which contained a triple-quoted string on one line, so if that passes I think we should be good.

File: crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs (hunk 8 above)
Comment (charliermarsh) [reply crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs:43]: `indent.find('\t')` returns the index of the starting _byte_ for the tab character, but some characters can require multiple bytes (think unicode) -- so `str.len()` in Rust doesn't always return what you might expect if coming from another programming language.

Comment (evanrittenhouse) [reply crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs:43]: Oh interesting, I didn't know that. I'll have to read up on it, thanks!
